File: Basketball Analytics/io.R
play_by_play  <- read.table('NBA Hackathon - Play by Play Data Sample (50 Games).txt', sep="", header=TRUE)
game_lineup   <- read.table('NBA Hackathon - Game Lineup Data Sample (50 Games).txt', sep="", header=TRUE)
event_codes   <- read.table('NBA Hackathon - Event Codes.txt', sep="", header=TRUE)


plus_minus    <- unique(game_lineup[c('Game_id', 'Person_id')])


play_by_play  <- play_by_play[order(play_by_play$Game_id,
                                    play_by_play$Period,
                                    -play_by_play$PC_Time,
                                    play_by_play$WC_Time,
                                    play_by_play$Event_Num),]
event_codes   <- event_codes[order(event_codes$Event_Msg_Type, 
                                   event_codes$Action_Type),]


for (game in unique(game_lineup$Game_id)) {
  
  lineups   <- game_lineup[game_lineup$Game_id == game,]        # df of starting lineups for each period in the game
  teams     <- unique(lineups$Team_id)                          # vector of the playing teams' IDs
  
  for (period in seq(1,4)) {  #CAN BE GREATER THAN 4 - find max period in the play by plays
    
    # df of lineups teams 1 and 2 (arbitrarily assigned) at the beginning of the current period
    current_lineup_1    <- factor(lineups[(lineups$Period == period) & (lineups$Team_id == teams[1]),]$Person_id)
    current_lineup_2    <- factor(lineups[(lineups$Period == period) & (lineups$Team_id == teams[2]),]$Person_id)
    print(current_lineup_1)
    # df of plays in the current period
    period_plays        <- play_by_play[(play_by_play$Game_id == game) & (play_by_play$Period == period),]

    for (play in 1:nrow(period_plays)) {
      event               <- period_plays[play, 'Event_Msg_Type']
        # 1   -   'Made shot'
        # 2   -   'Missed shot'
        # 3   -   'Free throw' (or 'No shot')
        # 4   -   'Rebound'
        # 5   -   'Turnover'
        # 6   -   'Foul'
        # 7   -   'Violation'
        # 8   -   'Substitution'
      action              <- period_plays[play, 'Action_Type']
      option1             <- period_plays[play, 'Option1']
      team                <- period_plays[play, 'Team_id']
      player1             <- period_plays[play, 'Person1']
      player2             <- period_plays[play, 'Person2']
      #player1team         <- ifelse(player1 %in% unique(game_lineup[game_lineup$Team_id == teams[1],]$Person_id), 1, ifelse(player1 %in% unique(game_lineup[game_lineup$Team_id == teams[2],]$Person_id), 2, 'Fuck'))
      #player2team         <- ifelse(player2 %in% unique(game_lineup[game_lineup$Team_id == teams[1],]$Person_id), 1, 2)
      
      if (event == 8) {                                       # 'Substitution'
        current_lineup_1 <- union(setdiff(current_lineup_1, player1), player2)
        
        print(player1)
        print(player2)
        print(current_lineup_1)
        
      } else if (event == 6) {                                # 'Foul'
        
      } else if (event == 3) {                                # 'Free throw'
        
      } else if (event == 1) {                                # 'Made shot'
        
      }
      
    }
  }
}

File: Basketball Analytics/io.py
from processing import read_data, sort_pbp, sort_ec, get_output
from parse import get_period_starters, sync_lineup, \
                  get_lag_events, assess_lag, clear_queue, update_stats, \
                  substitute, assess_ft_completion, assess_ft_completion_correction
import pandas as pd
import numpy as np

def parse_games(gl, pbp):
    stats_dict = {}

    games = gl.Game_id.unique()
    for game in games:
        game_plays = sort_pbp(pbp.loc[pbp.Game_id == game])
        game_period_starters = gl.loc[gl.Game_id == game]

        #print(game)
        stats_dict[game] = parse_periods(game_plays, game_period_starters, game, game_stats={})

    return stats_dict

def parse_periods(game_plays, game_period_starters, game, game_stats):
    for period in range(0, game_plays.Period.max() + 1):
        period_plays = sort_pbp(game_plays.loc[game_plays.Period == period])
        period_lineup = get_period_starters(game_period_starters, period)

        #print(period)
        game_stats = parse_plays(period_plays, period_lineup, game, game_stats)

    return game_stats

def parse_plays(period_plays, period_lineup, game, game_stats):
    period_plays = period_plays.reset_index(drop=True)
    queued_sub, queue_sub = clear_queue()

    for i, play in period_plays.iterrows():
        event = play.Event_Msg_Type
        action = play.Action_Type
        option1 = play.Option1
        team = play.Team_id
        other_team = [x for x in list(period_lineup.keys()) if x != team][0]
        person1 = play.Person1
        person2 = play.Person2

        game_stats = sync_lineup(period_lineup, game_stats)

        # free throw
        if (event == 3) & (action != 0):
            if queue_sub == True:
                game_stats = update_stats(game_stats,
                                          plus_team=period_lineup[team],
                                          minus_team=period_lineup[other_team],
                                          score=option1)

                period_lineup, action, queued_sub, queue_sub = assess_ft_completion(period_lineup, action,
                                                                                    queued_sub, queue_sub)

            elif queue_sub == False:
                # ad hoc solutions (more robust options?)
                period_lineup, game_stats = assess_ft_completion_correction(period_lineup, game_stats,
                                                                            team, other_team,
                                                                            game, action, option1)

        # made shot
        elif event == 1:
            game_stats = update_stats(game_stats,
                                      plus_team=period_lineup[team],
                                      minus_team=period_lineup[other_team],
                                      score=option1)

        # substitution
        elif event == 8:
            if queue_sub == False:
                period_lineup = substitute(period_lineup, person1, person2)
            else:
                queued_sub = substitute(period_lineup, person1, person2)

        # foul
        elif event == 6:
            lag_events = get_lag_events(i, period_plays)
            queue_sub = assess_lag(lag_events)

            #pause substitution until the completion of n free throws (if free throws awarded)

    return game_stats


def main():
    pbp, gl, ec = read_data({'pbp': 'NBA Hackathon - Play by Play Data Sample (50 Games).txt',
                             'gl':  'NBA Hackathon - Game Lineup Data Sample (50 Games).txt',
                             'ec':  'NBA Hackathon - Event Codes.txt'})
    ec = sort_ec(ec)
    #pbp = sort_pbp(pbp)

    stats_dict = parse_games(gl, pbp)
    output_df = get_output(stats_dict)

    output_df.to_csv('Crew_Q1_BBALL.csv', index=False)

if __name__ == '__main__':
    main()
